### ISSUE TITLE
Playlist: Move UP / DOWN does not react after removing songs 

### ISSUE THREAD

**Issue by Gero5** [mpogue2/SquareDesk#905]
Use case: I deleted several songs from my playlist. After that "move UP/DOWN" does not work as expected. It is not moving the song on the first click. I clicked it a second time and it works.
My first suspect: It seems like 'Move Up/DOWN' is counting/remembering the deleted songs.
(I have not the time to give that a second check at the moment)

**Comment by mpogue2** [comment]
@Gero5 What version are you running?  I am testing on 1.0.5 .

I have not been able to replicate this here.  If I have a playlist with say 10 items, 1 through 10, and I select and then delete item 5, that item does move to just past the bottom of the list (assuming that the songTable is sorted by playlist item number), because it is no longer on the playlist.  But, then I select item 4 OR item 5 (the new item 5!), and Move Up/Down works every time for me.  The playlist items are also all numbered correctly, as per the "#" column on the left.

Could you provide a more detailed example of how to reproduce this problem?  Thanks!

**Comment by Gero5** [comment]
version 1.0.5
I have a playlist with 22 items
- delete item # 17 
- right click on new item # 17 and select 'Move UP in playlist'
--> item # 17 does not move (bug)
- right click on new item # 17 and select 'Move UP in playlist'
- --> item moves
- right click on new item # 17 and select 'Move DOWN in playlist' 
- --> item does not move, but it changes the # to 18 and I have two entries with # 19 and two entries with # 20

**Comment by mpogue2** [comment]
@Gero5  Thanks for the details!  Wow that's weird.  I can't duplicate that here yet.

Do you have a filter set (e.g. "singing" in the Type search field)?  I see only singing calls here...
But, even when I do that kind of a filter, I can't get it to fail here!

The fact that you have two 19's and two 20's is already an indication that the playlist is messed up.  (I haven't been able to create that condition in the first place here yet...)

I'm speculating -- when you loaded the playlist, were there any errors?  I'm wondering if the playlist gets messed up when (and only when?) there are playlist load errors, for example if you changed the name of a song after the playlist was saved, and you later try to load that playlist.

**Comment by Gero5** [comment]
There are patter music items at # 1 - 10.

Probably more relevant than that: I am working on my **iMac at home** putting the playlist together. 
Music files and playlist files are stored on my **iCloud Drive.**
then I load that playlist in my **MacBook** to get ready for the clubnight.

Yesterday there were two singings that I bought new (Penny Lover, Walking By Myself). Due to Apple's lazy loading on iCloud drive the singing calls were not on the MacBook. As it seems they vanished from the playlist. 
I loaded those files later on during the clubnight. However, it does not seem like the playlist gets an update for that.
As I was not able to see the reloaded files in the playlist I moved them up one folder in the finder. (just to make the use case more interesting...)
Those music files were on the computer when I did the playlist modifications but in another folder. they are still missing in the displayed playlist. see screenshot above.
So maybe the missing/reloaded files are causing the problem?

here ist the complete playlist before
[2023-08-02 playlist before.csv](https://github.com/mpogue2/SquareDesk/files/12250631/2023-08-02.playlist.before.csv)

and the modified list after saving it
[2023-08-02 playlist after.csv](https://github.com/mpogue2/SquareDesk/files/12250699/2023-08-02.playlist.after.csv)

Penny Lover and Walking By Myself are erased from the list. My guess is that the file names are still in the internal memory although not visible. so the numbers on the display and the list in the internal memory might not fit to each other.

**Comment by mpogue2** [comment]
@Gero5 Thanks for the detailed explanation!  It does sound like the situation was that these files were in the playlist, but then on the OTHER computer those files were not found, and so they generated errors at playlist load time, and were correctly removed from the list (although remaining songs were not quite numbered correctly it seems, but the non-existent ones were definitely removed) from the playlist.

So, if this is the case, I can see two things going on here:

- BUG: when a playlist is loaded, and songs are not present, the numbering of the remaining songs gets messed up.  In your example, two item 19's and two item 20's should not have occurred.  This should never happen and it is a bug (I'll open this separately).

- FEATURE: be able to more gracefully handle the case where a third-party storage service (iCloud, Dropbox, Box.net) does not yet have the local copies of the files, and so playlist load fails (but could succeed later if the files actually were downloaded in the future).  

Right now, if a playlist references a file that does not exist, well, it really can't be on the playlist anymore, since it literally cannot be loaded.  There is an implied promise in the UX that the songTable lists all of the songs that can be loaded and played.  (At least, that's the way it works today.)

For this new feature to work, I think we'd need to invent a new concept.  Something like a "soft reference", if you are familiar with that terminology.  In this new concept use case, the song is listed in the songTable, but it might not actually exist.  Its name could be displayed in some other color, or with a strikethrough styling.  Then, when you TRY to load it, it could somehow fetch that file (assuming that Qt has some way to telling the OS to do that), give you a progress dialog while it downloads.  When it's done downloading, it could change the font styling to normal, and continue with the load.  

That assumes, of course, that the Internet is available for downloading, AND that the downloading proceeds quickly.  Otherwise, the user could be sitting there for a while.  AND, the user, getting frustrated, could try to access another (or several) un-downloaded songs, so we'd have a queue of them (or we could cancel the download of the first song when a second one is attempted).  If the user accesses an already-downloaded song while waiting, we might then want to abort the downloading of the un-downloaded song (so, essentially a one-item queue).

I use iCloud, Box.net, and Dropbox (Box.net is my preferred provider for the square dance music stuff).  In my testing just now on iCloud, telling MacOS to download an un-downloaded song put the icon the the cloud-without-down-arrow, so it thinks it is downloading, but the actual download is NOT happening.  If this were an MP3 file, we'd be in the "wait indefinitely" error use case described above.  

GetInfo in Finder shows (this is a PDF file) that the size is "1,455,440 bytes (4KB on disk)" for an un-downloaded PDF file, and "774,682 bytes (778 KB on disk)" for an already-downloaded local PDF file.

If these storage providers are all using OS functions, and if these are not accessible via Qt, then it's not likely that we can do anything to implement this feature, at least described the way I've described it above.

**Comment by Gero5** [comment]
There is an update event for the music list, I cannot say wether that is triggered after a file has been downloaded on iCloud.

However, as iCloud allows almost no user control afaik over how the system loads or unloads the files from the physical disc, I think iCloud is not the best choice for the music library. I see that it is erasing music files from my physical drive when they are not in use for a longer time. At club night I usually have no access to the internet. So wich music files I have or have not on my computer is rather unpredictable. I might give it a second thought.

I think with the Google drive sync app there is an option that one could mark folder to be physically mirrored on the computer. I miss that on iCloud drive.

**Comment by mpogue2** [comment]
@Gero5 I tried Dropbox, iCloud, and Box.net, and Box.net seems to work the best (I use the free version).  I still use Dropbox (free) and iCloud (paid) for other things, but for Square Dance Music, Box.net does the fewest surprising things with the files.

Yes, I think that missing files might have been the cause, but it's hard to say.  I tried to manually duplicate the load of a playlist with invalid entries, and I was not able to get it to fail.  But, I think it's a strong contender as root cause for the observed behavior.

Let's add in the new #909 feature, and after that see if this problem occurs again.

**Comment by Gero5** [comment]
> see if this problem occurs again.

I was not able to reproduce the bug with existing files.

**Comment by Gero5** [comment]
BTW: Does SquareDesk keep my choices for pitch, tempo and volume if the files are moved to another folder?

**Comment by mpogue2** [comment]
@gero No.  SquareDesk considers them to be two different files, if they are in two different folders.  This allows you to select different pitch/tempo/volume for the same file used as singer and patter, for example.  

This works both for files that are simply copied, and for files with aliases (e.g. you can use a singer as a patter by making an alias and moving that alias to the patter folder).

**Comment by Gero5** [comment]
@mpogue2 I have a number of music files where I changed pitch/tempo/volume.
Where are those informations stored by SquareDesk? 
Is it a readable (text) format? Can I access the file for instance with a text editor and replace the folder names after moving?

**Comment by mpogue2** [comment]
@Gero5 Yes, all this information is in the SQLite database.  You can manually edit that, if you are very careful, using a program like "DB Browser for SQLite".  

You might also be able to do your manual editing use case using File > Export Song Data, then edit, then File > Import Song Data.  But, I don't think anybody has ever tried that, so it might or might not work.  Nobody to my knowledge is actively using those menu items at this time.  If you want to try that, be sure to back up your music directory!  The SQLite DB (SquareDesk.sqlite3) is in the .squaredesk folder inside your music directory.  If you back up your music directory, it will also back up that DB.
